src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -970,13 +1375,196 @@ fn slice_pat_covered_by_constructor<'tcx>(
+/// change.
+/// Our solution, therefore, is to split the range constructor into subranges at every single point
+/// the group of intersecting patterns changes, which we can compute by converting each pattern to
+/// a range and recording its endpoints, then creating subranges between each consecutive pair of
+/// endpoints.
+/// And voilà! We're testing precisely those ranges that we need to, without any exhaustive matching
+/// on actual integers. The nice thing about this is that the number of subranges is linear in the
+/// number of rows in the matrix (i.e. the number of cases in the `match` statement), so we don't
+/// need to be worried about matching over gargantuan ranges.
+///
+/// Essentially, given the first column of a matrix representing ranges, looking like the following:
+///
+/// |------|  |----------| |-------|    ||
+///    |-------| |-------|            |----| ||
+///       |---------|
+///
+/// We truncate the ranges so that they lie inside each range constructor and then split them
+/// up into equivalence classes so the ranges are no longer overlapping:
+///
+/// |--|--|||-||||--||---|||-------|  |-|||| ||
+///
+/// The logic for determining how to split the ranges is a little involved: we need to make sure
+/// that we have a new range for each subrange for which a different set of rows coïncides, but
+/// essentially reduces to case analysis on the endpoints of the ranges.
+fn split_grouped_constructors<'p, 'a: 'p, 'tcx: 'a>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    ctors: Vec<Constructor<'tcx>>,
+    &Matrix(ref m): &Matrix<'p, 'tcx>,
+    ty: Ty<'tcx>,
+) -> Vec<Constructor<'tcx>> {
+    let mut split_ctors = Vec::with_capacity(ctors.len());
+
+    for ctor in ctors.into_iter() {
+        match ctor {
+            // For now, only ranges may denote groups of "subconstructors", so we only need to
+            // special-case constant ranges.
+            ConstantRange(..) if should_treat_range_exhaustively(tcx, &ctor) => {
+                // We only care about finding all the subranges within the range of the constructor
+                // range. Anything else is irrelevant, because it is guaranteed to result in
+                // `NotUseful`, which is the default case anyway, and can be ignored.
+                let ctor_range = IntRange::from_ctor(tcx, &ctor).unwrap();
+
+                // We're going to collect all the endpoints in the new pattern so we can create
+                // subranges between them.
+                // If there's a single point, we need to identify it as belonging
+                // to a length-1 range, so it can be treated as an individual
+                // constructor, rather than as an endpoint. To do this, we keep track of which
+                // endpoint a point corresponds to. Whenever a point corresponds to both a start
+                // and an end, then we create a unit range for it.
+                #[derive(PartialEq, Clone, Copy, Debug)]
+                enum Endpoint {
```

Comment:
A potential simplification:
I just noticed that range boundaries are always "in between" numbers - an `Endpoint::Start` is a boundary between `p-1` and `p`, and an `Endpoint::End` is between `p` and `p+1` (that means that an `Endpoint::End` is equivalent to an `Endpoint::Start` that immediately follows it). You could do something based on that.

Reply:
e.g., I think this is a correct implementation:

```Rust
/// Represents a border between 2 integers. Because of the "fencepost error",
/// there are be 2^128+1 such borders.
#[derive(Copy, Clone, PartialOrd, Ord, PartialEq, Eq)]
enum Border {
    JustBefore(u128),
    AfterU128Max
}

impl Border {
    fn before(n: u128) -> Self {
        Border::JustBefore(n)
    }

    fn after(n: u128) -> Self {
        match n.checked_add(1) {
            m => Border::JustBefore(m),
            None => Border::AfterU128Max
        }
    }

    fn range_borders(r: IntRange<'_>) -> [Self; 2] {
        let (r_lo, r_hi) = r.range.into_inner();
        vec![Border::before(r_lo), Border::after(r_hi)]
    }

    // return the integers between `self` and `to` if `self < to`, `None` otherwise.
    fn range_to(self, to: Self, ty: Ty<'tcx>) -> Option<IntRange<'tcx>> {
        match (self, to) {
            (Border::JustBefore(n), Border::JustBefore(m)) => {
                IntRange::from_interval(ty, n, m, RangeEnd::Excluded)
            }
            (Border::JustBefore(n), Border::AfterU128Max) => {
                IntRange::from_interval(ty, n, u128::MAX, RangeEnd::Included)
            }
            (Border::AfterU128Max, _) => None
        }
    }
}

// `borders` is the set of borders between equivalence classes - each equivalence
// class is between 2 borders.
let row_borders = m.iter()
    .flat_map(|row| IntRange::from_pat(tcx, row[0]))
    .flat_map(|range| ctor_range.intersection(&r))
    .flat_map(|range| Border::range_borders(range));
let range_borders = Border::range_borders(ctor_range);
let mut borders: Vec<Border> = row_borders.chain(range_borders).collect();
borders.sort();

let ranges = borders.windows(2).map(|window| {
    window[0].range_to(window[1]);
})
```

Reply:
I do think this is cleaner :) The logic looks good to me too — I'm going to give it a go.